src/cascadia/TerminalConnection/AzureConnection.cpp

```diff
@@ -0,0 +1,478 @@
+    {
+        if (!_connected || !_termConnected)
+        {
+            _initialRows = rows;
+            _initialCols = columns;
+        }
+        else if (!_closing.load())
+        {
+            //Initialize client
+            http_client terminalClient(_cloudShellUri);
+
+            //Initialize the request
+            http_request terminalRequest(L"POST");
+            terminalRequest.set_request_uri(L"terminals/" + _terminalID + L"/size?cols=" + std::to_wstring(columns) + L"&rows=" + std::to_wstring(rows) + L"&version=2019-01-01");
+            terminalRequest.headers().add(L"Accept", L"application/json");
+            terminalRequest.headers().add(L"Content-Type", L"application/json");
+            terminalRequest.headers().add(L"Authorization", L"Bearer " + _accessToken);
+            terminalRequest.headers().add(L"User-Agent", L"Terminal");
+            terminalRequest.set_body(json::value(L""));
+
+            //Send the request
+            _requestHelper(terminalClient, terminalRequest);
+        }
+    }
+
+    void AzureConnection::Close()
+    {
+        if (!_connected)
+        {
+            return;
+        }
+
+        if (!_closing.exchange(true))
+        {
+            _canProceed.notify_all();
+            if (_termConnected)
+            {
+                // close the websocket connection
+                pplx::task<void> closed = _cloudShellSocket.close();
+                closed.wait();
+            }
+
+            // Tear down our output thread
+            WaitForSingleObject(_hOutputThread.get(), INFINITE);
+            _hOutputThread.reset();
+        }
+    }
+
+    DWORD WINAPI AzureConnection::StaticOutputThreadProc(LPVOID lpParameter)
+    {
+        AzureConnection* const pInstance = (AzureConnection*)lpParameter;
+        return pInstance->_OutputThread();
+    }
+
+    DWORD AzureConnection::_OutputThread()
+    {
+        //Azure authentication starts here:
+        //step 1a: initiate device code flow
+        json::value deviceCodeResponse = _getDeviceCode();
+
+        //step 1b: print the message and store the device code, polling interval and expiry
+        utility::string_t message = deviceCodeResponse.at(L"message").as_string();
+        _outputHandlers(message + L" This code will expire in 15 minutes.\r\n");
+        utility::string_t devCode = deviceCodeResponse.at(L"device_code").as_string();
+        int pollInterval = std::stoi(deviceCodeResponse.at(L"interval").as_string());
+        int expiresIn = std::stoi(deviceCodeResponse.at(L"expires_in").as_string());
+
+        //step 2: wait for user authentication and obtain the access/refresh tokens
+        json::value authenticatedResponse;
+        try
+        {
+            authenticatedResponse = _waitForUser(devCode, pollInterval, expiresIn);
+        }
+        catch (...)
+        {
+            _outputHandlers(winrt::to_hstring("Exit."));
+            return -1;

```

Comment:
Consider returning known HRESULT failure codes from this thread like E_FAIL. You can find a ton of well-known codes by finding E_FAIL, S_OK, and friends in the windows headers. (Go to definition on one of them)